clang/lib/Sema/SemaDecl.cpp

```diff
@@ -11170,6 +11170,12 @@ Sema::ActOnFunctionDeclarator(Scope *S, Declarator &D, DeclContext *DC,
   if (getLangOpts().OpenACC)
     OpenACC().ActOnFunctionDeclarator(NewFD);
 
+  // Handle free functions.
+  if (LangOpts.SYCLIsDevice && !NewFD->isDependentContext() &&
+      !D.isRedeclaration() &&
```

Comment:
This check is probably unnecessary if we make FreeFunctionDeclarations a Set and put canonical declarations there.
```suggestion
```